src/manifest.js

```diff
@@ -0,0 +1,157 @@
+      if (!dateDescriptor.match(/^\d{4}-\d{2}-\d{2}$/)) {
+        return false;
+      }
+      const availableDates = (this.data?.versions || []).map((v) => v.date);
+      if (!availableDates.length) {
+        return false;
+      }
+      let closest = availableDates[0];
+      if (closest < dateDescriptor) {
+        return LATEST;
+      }
+      for (const date of availableDates) {
+        if (date >= dateDescriptor) {
+          closest = date;
+        } else {
+          break;
+        }
+      }
+      return closest;
+    }
+
+
+    /**
+     * Given a YYYY-MM-DD string, return the available subresources and their corresponding URLs
+     */
+    subresourceUrls(date) {
+      for (const version of this.data?.versions || []) {
+        if (version.date===date) {
+          return version.fileUrls;
+        }
+      }
+      throw new Error(`Attempted to access a specific date which doesn't exist in the manifest: ${date}`)
+    }
+
+  }
+
+  return {updateResourceVersions, ResourceVersions}
+
+})();
+
+
+
+/**
+ * Updates provided data in-place by fetching the file on disk.
+ *
+ * TODO - for S3 fetches we should examine the eTag to prevent unnecessary data
+ * fetching / server compute when the underlying data hasn't changed. This will
+ * allow us to call the update method more frequently.
+ * @param {*} data 
+ */
+async function _update(resources, eTag) {
+  let newResources;
+  if (process.env.LOCAL_MANIFEST) {
+    utils.verbose(`Updating available resources via ${process.env.LOCAL_MANIFEST}`);
+    newResources = JSON.parse(
+      await (new Promise((resolve, reject) => {
+        fs.readFile(process.env.LOCAL_MANIFEST, (err, data) => {
+          /* Don't catch any error here as this is only used for development and
+          we should exit loudly */
+          if (err) reject(err);
+          resolve(data);
+        })
+      })).then((contents) =>
+        process.env.LOCAL_MANIFEST.endsWith('.gz') ? gunzip(contents) : contents
+      )
+    )
+  } else {
+    const [Bucket, Key] = ['nextstrain-inventories', 'resources.json.gz'];
+    utils.verbose(`Updating available resources by fetching s3://${Bucket}/${Key}`);
+    const S3 = new AWS.S3();
+    try {
+      const newETag = await (new Promise((resolve, reject) => {
+        S3.headObject({Bucket, Key}, (err, data) => {
+          if (err) return reject(err);
+          resolve(data.ETag);
+        })
+      }));
+      if (newETag === eTag.value) {
+        utils.verbose("Skipping available resource update as eTag hasn't changed");
+        return;
+      }
+      newResources = JSON.parse(
+        await (new Promise((resolve, reject) => {
+          S3.getObject({Bucket, Key}, (err, data) => {
+            if (err) return reject(err);
+            /* NOTE - the Body property of the response is a Buffer in S3 SDK v2
+            (which is the current version in the nextstrain.org codebase) but
+            this has changed to a readable stream in SDK v3 */
```

Comment:
Yeah. My view is to separate out the upgrading of deps from new functionality _unless_ the new functionality would benefit from the updated dependency. In this case I don't think v3 offers any significant improvements for what I'm doing here.

Reply:
v2 also has a promise-based interface so I've used that which makes things nicer IMO. For instance, getting the eTag is a one-liner:
```js
const newETag = (await S3.headObject({Bucket, Key}).promise()).ETag;
```

Reply:
Nod. Yeah, that's nicer.

FWIW, I updated the only use of v2 on `master` to v3 in another PR (because we needed a v3-only feature): https://github.com/nextstrain/nextstrain.org/pull/733/commits/4df463b1622723d36bf9f7c311f94a9358bb155e